### DISCUSSION TITLE
Would it be possible to get the last mouse position in Cosmos?

### DISCUSSION THREAD

**Discussion by elkaamee326** [CosmosOS/Cosmos#1766] [Q&A]
I know that `MouseManager.X` and `Y` updates everytime the mouse moves (when you invoke it during `Run();`). How could I capture the last position that the mouse was in and use those values when drawing a line. This would be useful because I am currently working on a Paint Widget but I don't know how to get the last position. This is my current code:

```CSharp 
using System;
using System.Drawing;
using Cosmos.System.Graphics;
using System.Collections.Generic;
using System.Text;

using Cosmos.System;
namespace CosmosKernel1
{
    class Paint : App
    {
        Pen PaintBrush;
        int? lastPosX, lastPosY = null;
        public Paint(uint width, uint height, uint x = 0, uint y = 0) : base(width, height, x, y)
        {
            name = "Paint";
        }

        public override void _Update()
        {
            bool Collide = CosmosKernel1.Utils.Utils.DoesMouseCollideWithArea((int)baseX, (int)baseY, (int)baseWidth, (int)baseHeight);

            if (Kernel.Pressed)
            {
                if (Collide == true)
                {
                    // If inside of the window bounds...
                    PaintBrush.Color = Color.Blue;
                    if (PaintBrush.Color != Color.Empty)
                    {
                        Kernel.vMWareSVGAII.DoubleBuffer_DrawLine((uint)PaintBrush.Color.ToArgb(), (int)(lastPosX ?? (int?)MouseManager.X), (int)(lastPosY ?? (int?)MouseManager.Y), (int)MouseManager.X, (int)MouseManager.Y);
                        lastPosX = (int?)MouseManager.X;
                        lastPosY = (int?)MouseManager.Y;
                    }
                }
            }
            else 
            {
                lastPosX = null;
                lastPosY = null;
                PaintBrush.Color = Color.Empty;
            }
        }
    }
}
```

https://user-images.githubusercontent.com/69808082/117737999-08da5000-b1c9-11eb-832a-ef8951ac7d35.mp4
(The lag at the end and throughout the video was just produced from filming... Not the actual code... (Almost 0 lag in the code...))

Also, how could I draw those pixels onto the screen because when I run the code, it draws a line which follows my cursor position (this is a lazer) but how can I draw lines onto the screen (specifically onto the form)?
Thanks :)

**Comment by elkaamee326** [comment]
Does anyone know how to make a working paint app that runs on Cosmos? Or a reference of a project that might work? Thanks :)

**Comment by AnErrupTion** [comment]
You can simply store two variables, `LastX` and `LastY`, then do your stuff and update them to the current X and Y respectively, then use them later on as the X and Y change.

**Reply by elkaamee326** [reply]
Could you please give a code example? Also, thanks for the reply :D Highly appreciated.

**Reply by AnErrupTion** [reply]
```
int LastX, LastY;

// mouse stuff with X and Y

LastX = Mouse.X;
LastY = Mouse.Y;
```

**Comment by ghost** [comment]
I like events, they do the job.
